File: Sources/AislopdeskVideoHost/RecoveryIDRPolicy.swift
import AislopdeskVideoProtocol

/// DELIVERY-KEYED recovery-IDR admission policy (component 2, 2026-06-11) — replaces the
/// capturer's sent-keyed F1 cooldown (`AISLOPDESK_MIN_IDR_MS`, 500 ms) as the single authority on
/// whether a client recovery request may force a real IDR.
///
/// THE BUG THIS FIXES (the ranked-#1 hitch): the F1 gate keyed the cooldown on keyframe SEND
/// time. When BOTH kfDup copies of a recovery IDR were lost (burst), the client's 2·RTT
/// escalation re-requested every ~2·RTT — and EVERY request landing inside the 500 ms window was
/// suppressed (the host kept shipping P-frames the broken client could not use). Worst case
/// ~600 ms of freeze, C-dominated and RTT-independent. Delivery-keying removes the C term: a
/// request that carries `lastDecodedFrameID < newest sent keyframe` past the in-flight grace
/// PROVES that keyframe is a casualty ⇒ grant immediately (the casualty bypass).
///
/// Decision table (`r` = request's lastDecoded, `K` = newest sent keyframe):
///  - r ≥ K                       ⇒ the request itself proves K delivered + reports a genuinely
///                                  new post-K loss ⇒ grant (token-gated).
///  - r < K, age(K) <  grace      ⇒ request plausibly crossed K in flight ⇒ suppress; if K was
///                                  lost the client re-escalates 2·RTT later into the next row.
///  - r < K, age(K) ≥  grace      ⇒ K presumed a casualty ⇒ THE BYPASS: grant immediately.
///  - r < a keyframe the client decode-ACKED ⇒ stale request from before the client's own
///                                  re-anchor ⇒ suppress at zero cost regardless of age.
///  - token bucket (cap 2, refill 1/500 ms) caps everything that reaches "grant": sustained
///    rate identical to the old F1 (≤2/s), burst of 2 so the casualty-bypass second IDR is
///    never blocked.
///
/// PURE + WALL-CLOCK-ONLY: all time injected as `Double` seconds (the session's `systemUptime`
/// domain), zero frame counting — immune to FPS-governor cadence changes. No Apple imports
/// (`UInt32.distanceWrapped` comes from AislopdeskVideoProtocol). Headlessly unit-testable like
/// ``LiveCongestionController``.
public struct RecoveryIDRPolicy: Sendable, Equatable {
    public struct Config: Sendable, Equatable {
        /// In-flight grace = `graceFraction × smoothedRTT`, clamped to [floor, ceil]. A crossing
        /// request arrives ≤ RTT/2 + jitter after the keyframe send; 0.75×RTT adds ~50% jitter
        /// margin (the measured path jitters RTT 10-59 ms).
        public var graceFraction: Double = 0.75
        /// Covers the rtt-unknown bootstrap (smoothedRTT = 0 before the first netstats fold).
        public var graceFloorSeconds: Double = 0.040
        /// = kfDupMinInterval: beyond it the kfDup second copy has also long been sent, so
        /// further suppression only adds freeze.
        public var graceCeilSeconds: Double = 0.250
        /// Burst allowance: exactly one ordinary grant + one casualty-bypass grant back-to-back
        /// (recovery IDRs are compact + kfDup-doubled ⇒ 2 grants ≈ 4 wire copies in <500 ms —
        /// bounded; 3+ would re-open the F1 storm).
        public var bucketCapacity: Double = 2.0
        /// 1 token / 500 ms sustained — preserves the old F1 spacing ceiling exactly.
        public var refillTokensPerSecond: Double = 2.0
        /// A granted-but-unserviced latch suppresses duplicates until this expires. Sized above
        /// the worst legitimate latch-service path: a freshly-quiet window waits the
        /// StaticIDRDecider quietWindow (1.0 s) + timer tick (0.25 s) + margin. Prevents both
        /// premature double-grants and a permanent wedge if capture dies.
        public var grantPendingTimeout: Double = 1.5
        /// Keyframes are rare (recovery + static-crisp + first-frame; motion heartbeat default
        /// OFF) — 4 covers every keyframe plausibly in flight within one ack round-trip.
        public var keyframeRingCapacity: Int = 4
        public init() {}
    }

    /// One sent keyframe (a tuple won't synthesize Equatable).
    public struct SentKeyframe: Sendable, Equatable {
        public let id: UInt32
        public let at: Double
        public init(id: UInt32, at: Double) {
            self.id = id
            self.at = at
        }
    }

    public enum Verdict: Equatable, Sendable {
        case grant
        /// An IDR grant is already latched and unexpired — the duplicate-request absorber.
        case suppressGrantPending
        /// The request provably predates a keyframe the client DECODED (acked) — zero-cost
        /// suppression regardless of age.
        case suppressStale
        /// The newest sent keyframe plausibly is still in flight to the client.
        case suppressInFlight
        /// Token bucket empty — the storm cap.
        case suppressRateLimited
    }

    public let config: Config
    /// Newest-last ring of recently-sent keyframes, capped at `keyframeRingCapacity`.
    private var recentKeyframes: [SentKeyframe] = []
    /// Newest SENT-keyframe id the client decode-ACKED (ring-matched — an LTR-P ack never
    /// masquerades as keyframe delivery).
    private var deliveredKeyframeID: UInt32?
    private var tokens: Double
    private var lastRefillAt: Double?
    /// Time of the last `.grant` not yet serviced by a `noteKeyframeSent` (nil = none pending).
    private var grantedAt: Double?

    public init(config: Config = Config()) {
        self.config = config
        tokens = config.bucketCapacity
    }

    /// Read-only token level (observability/tests — proves suppress* verdicts spend nothing).
    public var availableTokens: Double { tokens }

    /// Called from `onEncodedFrame` for EVERY keyframe handed to the wire (recovery, first-frame,
    /// static-crisp, heartbeat) with `packetizer.peekNextFrameID` read BEFORE packetize.
    public mutating func noteKeyframeSent(frameID: UInt32, now: Double) {
        recentKeyframes.append(SentKeyframe(id: frameID, at: now))
        if recentKeyframes.count > config.keyframeRingCapacity { recentKeyframes.removeFirst() }
        grantedAt = nil // a keyframe went out: any pending grant is serviced
    }

    /// Called from the `.ack` fold. Idempotent; only ids matching a ring entry count (an LTR-P
    /// ack must not masquerade as keyframe delivery). Wrap-aware keep-newest.
    public mutating func noteKeyframeDelivered(frameID: UInt32) {
        guard recentKeyframes.contains(where: { $0.id == frameID }) else { return }
        if let delivered = deliveredKeyframeID, frameID.distanceWrapped(from: delivered) <= 0 { return }
        deliveredKeyframeID = frameID
    }

    /// THE admission decision for one IDR-issuing recovery request.
    /// `clientLastDecoded == nil` ⇔ wire sentinel "nothing decoded yet" (treated as maximally
    /// behind — the connect-time first-IDR-loss case rides the same bypass).
    public mutating func decide(now: Double, clientLastDecoded: UInt32?, smoothedRTTSeconds: Double) -> Verdict {
        refill(now: now)
        if let granted = grantedAt, now - granted < config.grantPendingTimeout {
            return .suppressGrantPending
        }
        if let delivered = deliveredKeyframeID, let request = clientLastDecoded,
           request.distanceWrapped(from: delivered) < 0
        {
            // Exact, not heuristic: the client's lastDecoded is monotonic, so a request older
            // than a keyframe it ACKED was composed before that keyframe decoded — stale.
            return .suppressStale
        }
        if let newest = recentKeyframes.last {
            // nil lastDecoded (nothing decoded yet) is maximally behind by definition.
            let clientBehind = clientLastDecoded.map { $0.distanceWrapped(from: newest.id) < 0 } ?? true
            if clientBehind, now - newest.at < grace(rtt: smoothedRTTSeconds) {
                return .suppressInFlight
            }
        }
        guard tokens >= 1.0 else { return .suppressRateLimited }
        tokens -= 1.0
        grantedAt = now
        return .grant
    }

    /// In-flight grace window for the given smoothed RTT: clamp(graceFraction × rtt, floor, ceil).
    public func grace(rtt: Double) -> Double {
        min(config.graceCeilSeconds, max(config.graceFloorSeconds, config.graceFraction * rtt))
    }

    private mutating func refill(now: Double) {
        defer { lastRefillAt = now }
        guard let last = lastRefillAt, now > last else { return }
        tokens = min(config.bucketCapacity, tokens + (now - last) * config.refillTokensPerSecond)
    }
}

File: rust/aislopdesk-core/src/recovery_idr_policy.rs
//! Delivery-keyed recovery-IDR admission policy — a port of Swift `RecoveryIDRPolicy`.
//!
//! The single authority on whether a client recovery request may force a real IDR. The legacy
//! gate keyed its cooldown on keyframe SEND time: when both kfDup copies of a recovery IDR were
//! lost, the client's 2·RTT escalation re-requested every ~2·RTT and every request inside the
//! 500 ms window was suppressed → ~600 ms of freeze, RTT-independent. Delivery-keying removes that:
//! a request carrying `last_decoded < newest sent keyframe` past the in-flight grace PROVES that
//! keyframe is a casualty ⇒ grant immediately (the casualty bypass).
//!
//! Decision table (`r` = request's last-decoded, `K` = newest sent keyframe):
//!  - r ≥ K ⇒ the request proves K delivered + a new post-K loss ⇒ grant (token-gated).
//!  - r < K, age(K) < grace ⇒ K plausibly crossed in flight ⇒ suppress.
//!  - r < K, age(K) ≥ grace ⇒ K presumed a casualty ⇒ THE BYPASS: grant.
//!  - r < a keyframe the client decode-ACKED ⇒ stale ⇒ suppress at zero cost regardless of age.
//!  - a token bucket (cap 2, refill 1/500 ms) caps everything that reaches "grant".
//!
//! Pure + wall-clock-only: all time injected as `f64` seconds (the session's `systemUptime`
//! domain), zero frame counting — immune to FPS-governor cadence changes.

use crate::seq::distance_wrapped;

/// Tunables for [`RecoveryIdrPolicy`]. f64 fields, so [`PartialEq`] only.
#[derive(Debug, Clone, Copy, PartialEq)]
pub struct Config {
    /// In-flight grace = `grace_fraction × smoothed_rtt`, clamped to `[floor, ceil]`.
    pub grace_fraction: f64,
    /// Covers the rtt-unknown bootstrap (smoothed RTT = 0 before the first netstats fold).
    pub grace_floor_seconds: f64,
    /// Beyond it the kfDup second copy has also long been sent, so further suppression only adds
    /// freeze (= `kfDupMinInterval`).
    pub grace_ceil_seconds: f64,
    /// Burst allowance: one ordinary grant + one casualty-bypass grant back-to-back.
    pub bucket_capacity: f64,
    /// Sustained refill rate (1 token / 500 ms preserves the old spacing ceiling).
    pub refill_tokens_per_second: f64,
    /// A granted-but-unserviced latch suppresses duplicates until this expires.
    pub grant_pending_timeout: f64,
    /// Newest-last ring size for recently-sent keyframes.
    pub keyframe_ring_capacity: usize,
}

impl Default for Config {
    fn default() -> Self {
        Self {
            grace_fraction: 0.75,
            grace_floor_seconds: 0.040,
            grace_ceil_seconds: 0.250,
            bucket_capacity: 2.0,
            refill_tokens_per_second: 2.0,
            grant_pending_timeout: 1.5,
            keyframe_ring_capacity: 4,
        }
    }
}

/// One sent keyframe (id + send time in the `systemUptime` domain).
#[derive(Debug, Clone, Copy, PartialEq)]
pub struct SentKeyframe {
    /// The keyframe's frame id.
    pub id: u32,
    /// The send time (seconds).
    pub at: f64,
}

/// The admission verdict for one IDR-issuing recovery request.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Verdict {
    /// Issue the real IDR.
    Grant,
    /// An IDR grant is already latched and unexpired — the duplicate-request absorber.
    SuppressGrantPending,
    /// The request provably predates a keyframe the client DECODED (acked) — zero-cost suppression.
    SuppressStale,
    /// The newest sent keyframe plausibly is still in flight to the client.
    SuppressInFlight,
    /// Token bucket empty — the storm cap.
    SuppressRateLimited,
}

/// Delivery-keyed admission policy for recovery IDRs.
#[derive(Debug, Clone, PartialEq)]
pub struct RecoveryIdrPolicy {
    config: Config,
    recent_keyframes: Vec<SentKeyframe>,
    delivered_keyframe_id: Option<u32>,
    tokens: f64,
    last_refill_at: Option<f64>,
    granted_at: Option<f64>,
}

impl Default for RecoveryIdrPolicy {
    fn default() -> Self {
        Self::new(Config::default())
    }
}

impl RecoveryIdrPolicy {
    /// Builds a policy with the given config; the token bucket starts full.
    #[must_use]
    pub const fn new(config: Config) -> Self {
        let tokens = config.bucket_capacity;
        Self {
            config,
            recent_keyframes: Vec::new(),
            delivered_keyframe_id: None,
            tokens,
            last_refill_at: None,
            granted_at: None,
        }
    }

    /// The active config.
    #[must_use]
    pub const fn config(&self) -> &Config {
        &self.config
    }

    /// Read-only token level (proves `suppress*` verdicts spend nothing).
    #[must_use]
    pub const fn available_tokens(&self) -> f64 {
        self.tokens
    }

    /// Called for EVERY keyframe handed to the wire (recovery, first-frame, static-crisp,
    /// heartbeat) with the next frame id read BEFORE packetize.
    pub fn note_keyframe_sent(&mut self, frame_id: u32, now: f64) {
        self.recent_keyframes.push(SentKeyframe {
            id: frame_id,
            at: now,
        });
        if self.recent_keyframes.len() > self.config.keyframe_ring_capacity {
            self.recent_keyframes.remove(0);
        }
        self.granted_at = None; // a keyframe went out: any pending grant is serviced
    }

    /// Called from the `.ack` fold. Idempotent; only ids matching a ring entry count (an LTR-P ack
    /// must not masquerade as keyframe delivery). Wrap-aware keep-newest.
    pub fn note_keyframe_delivered(&mut self, frame_id: u32) {
        if !self.recent_keyframes.iter().any(|k| k.id == frame_id) {
            return;
        }
        if let Some(delivered) = self.delivered_keyframe_id {
            if distance_wrapped(frame_id, delivered) <= 0 {
                return;
            }
        }
        self.delivered_keyframe_id = Some(frame_id);
    }

    /// THE admission decision for one IDR-issuing recovery request. `client_last_decoded == None`
    /// ⇔ the wire sentinel "nothing decoded yet" (treated as maximally behind).
    pub fn decide(
        &mut self,
        now: f64,
        client_last_decoded: Option<u32>,
        smoothed_rtt_seconds: f64,
    ) -> Verdict {
        self.refill(now);
        if let Some(granted) = self.granted_at {
            if now - granted < self.config.grant_pending_timeout {
                return Verdict::SuppressGrantPending;
            }
        }
        if let (Some(delivered), Some(request)) = (self.delivered_keyframe_id, client_last_decoded)
        {
            if distance_wrapped(request, delivered) < 0 {
                // The client's last-decoded is monotonic, so a request older than a keyframe it
                // ACKED was composed before that keyframe decoded — stale.
                return Verdict::SuppressStale;
            }
        }
        if let Some(newest) = self.recent_keyframes.last() {
            // None last-decoded (nothing decoded yet) is maximally behind by definition.
            let client_behind =
                client_last_decoded.is_none_or(|r| distance_wrapped(r, newest.id) < 0);
            if client_behind && now - newest.at < self.grace(smoothed_rtt_seconds) {
                return Verdict::SuppressInFlight;
            }
        }
        if self.tokens < 1.0 {
            return Verdict::SuppressRateLimited;
        }
        self.tokens -= 1.0;
        self.granted_at = Some(now);
        Verdict::Grant
    }

    /// In-flight grace window for the given smoothed RTT: `clamp(grace_fraction × rtt, floor, ceil)`.
    #[must_use]
    pub fn grace(&self, rtt: f64) -> f64 {
        (self.config.grace_fraction * rtt)
            .max(self.config.grace_floor_seconds)
            .min(self.config.grace_ceil_seconds)
    }

    fn refill(&mut self, now: f64) {
        // Mirrors the Swift `defer { lastRefillAt = now }`: the stamp advances on EVERY call, the
        // refill only when a strictly-later `now` has a prior stamp to measure against.
        if let Some(last) = self.last_refill_at {
            if now > last {
                self.tokens = self
                    .config
                    .bucket_capacity
                    .min(self.tokens + (now - last) * self.config.refill_tokens_per_second);
            }
        }
        self.last_refill_at = Some(now);
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn approx(a: f64, b: f64) {
        assert!((a - b).abs() < 1e-9, "{a} !~= {b}");
    }

    #[test]
    fn grant_when_no_recent_keyframe() {
        let mut policy = RecoveryIdrPolicy::default();
        approx(policy.available_tokens(), 2.0);
        assert_eq!(policy.decide(10.0, None, 0.05), Verdict::Grant);
        approx(policy.available_tokens(), 1.0);
    }

    #[test]
    fn casualty_bypass_after_grace() {
        let mut policy = RecoveryIdrPolicy::default();
        policy.note_keyframe_sent(100, 5.0);
        assert_eq!(policy.decide(5.2, Some(99), 0.05), Verdict::Grant);
    }

    #[test]
    fn suppress_in_flight_within_grace() {
        let mut policy = RecoveryIdrPolicy::default();
        policy.note_keyframe_sent(100, 5.0);
        let before = policy.available_tokens();
        assert_eq!(
            policy.decide(5.02, Some(99), 0.05),
            Verdict::SuppressInFlight
        );
        approx(policy.available_tokens(), before);
    }

    #[test]
    fn request_proves_delivery_grants() {
        let mut policy = RecoveryIdrPolicy::default();
        policy.note_keyframe_sent(100, 5.0);
        assert_eq!(policy.decide(5.01, Some(100), 0.05), Verdict::Grant);
        policy.note_keyframe_sent(101, 5.02);
        assert_eq!(policy.decide(5.03, Some(105), 0.05), Verdict::Grant);
    }

    #[test]
    fn suppress_stale_after_delivered_ack() {
        let mut policy = RecoveryIdrPolicy::default();
        policy.note_keyframe_sent(100, 5.0);
        policy.note_keyframe_delivered(100);
        let before = policy.available_tokens();
        assert_eq!(policy.decide(9.0, Some(99), 0.05), Verdict::SuppressStale);
        approx(policy.available_tokens(), before);
    }

    #[test]
    fn delivered_ack_ignored_unless_ring_match() {
        let mut policy = RecoveryIdrPolicy::default();
        policy.note_keyframe_sent(100, 5.0);
        policy.note_keyframe_delivered(555);
        assert_eq!(policy.decide(5.3, Some(99), 0.05), Verdict::Grant);
    }

    #[test]
    fn token_bucket_caps_burst_at_two() {
        let mut policy = RecoveryIdrPolicy::default();
        assert_eq!(policy.decide(10.0, None, 0.05), Verdict::Grant);
        policy.note_keyframe_sent(1, 10.01);
        assert_eq!(policy.decide(10.1, Some(0), 0.05), Verdict::Grant);
        policy.note_keyframe_sent(2, 10.11);
        assert_eq!(
            policy.decide(10.2, Some(0), 0.05),
            Verdict::SuppressRateLimited
        );
        assert_eq!(policy.decide(10.75, Some(0), 0.05), Verdict::Grant);
    }

    #[test]
    fn suppress_grant_pending_until_keyframe_sent_or_timeout() {
        let mut policy = RecoveryIdrPolicy::default();
        assert_eq!(policy.decide(10.0, None, 0.05), Verdict::Grant);
        assert_eq!(
            policy.decide(10.1, None, 0.05),
            Verdict::SuppressGrantPending
        );
        assert_eq!(
            policy.decide(11.0, None, 0.05),
            Verdict::SuppressGrantPending
        );
        policy.note_keyframe_sent(50, 11.1);
        assert_eq!(policy.decide(11.11, None, 0.05), Verdict::SuppressInFlight);

        let mut wedged = RecoveryIdrPolicy::default();
        assert_eq!(wedged.decide(20.0, None, 0.05), Verdict::Grant);
        assert_eq!(
            wedged.decide(21.4, None, 0.05),
            Verdict::SuppressGrantPending
        );
        assert_eq!(wedged.decide(21.6, None, 0.05), Verdict::Grant);
    }

    #[test]
    fn nil_last_decoded_treated_as_behind() {
        let mut policy = RecoveryIdrPolicy::default();
        policy.note_keyframe_sent(0, 5.0);
        assert_eq!(policy.decide(5.01, None, 0.05), Verdict::SuppressInFlight);
        assert_eq!(policy.decide(5.2, None, 0.05), Verdict::Grant);
    }

    #[test]
    fn wrap_aware_ids() {
        let mut policy = RecoveryIdrPolicy::default();
        let near_max: u32 = u32::MAX - 1;
        policy.note_keyframe_sent(near_max, 5.0);
        assert_eq!(policy.decide(5.01, Some(3), 0.05), Verdict::Grant);
        policy.note_keyframe_sent(2, 5.02);
        policy.note_keyframe_delivered(near_max);
        policy.note_keyframe_delivered(2);
        assert_eq!(
            policy.decide(9.0, Some(u32::MAX), 0.05),
            Verdict::SuppressStale
        );
        policy.note_keyframe_delivered(near_max);
        assert_eq!(
            policy.decide(9.5, Some(u32::MAX), 0.05),
            Verdict::SuppressStale
        );
    }

    #[test]
    fn grace_clamp() {
        let policy = RecoveryIdrPolicy::default();
        approx(policy.grace(0.0), 0.040);
        approx(policy.grace(0.059), 0.044_25);
        approx(policy.grace(1.0), 0.250);
    }

    #[test]
    fn wall_clock_only() {
        fn run(extra_probes: usize) -> Vec<Verdict> {
            let mut policy = RecoveryIdrPolicy::default();
            let mut verdicts = Vec::new();
            policy.note_keyframe_sent(10, 1.0);
            verdicts.push(policy.decide(1.02, Some(9), 0.05));
            for _ in 0..extra_probes {
                let _ = policy.decide(1.02, Some(9), 0.05);
            }
            verdicts.push(policy.decide(1.2, Some(9), 0.05));
            policy.note_keyframe_sent(11, 1.21);
            verdicts.push(policy.decide(1.3, Some(9), 0.05));
            policy.note_keyframe_sent(12, 1.31);
            verdicts.push(policy.decide(1.4, Some(9), 0.05));
            verdicts
        }
        assert_eq!(run(0), run(7));
        assert_eq!(
            run(0),
            vec![
                Verdict::SuppressInFlight,
                Verdict::Grant,
                Verdict::Grant,
                Verdict::SuppressRateLimited,
            ]
        );
    }

    #[test]
    fn keyframe_ring_evicts_oldest() {
        let mut policy = RecoveryIdrPolicy::default();
        for id in [1u32, 2, 3, 4, 5] {
            policy.note_keyframe_sent(id, 5.0 + f64::from(id) * 0.001);
        }
        policy.note_keyframe_delivered(1);
        assert_eq!(policy.decide(6.0, Some(0), 0.05), Verdict::Grant);
    }
}
